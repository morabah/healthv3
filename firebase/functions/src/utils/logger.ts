/**
 * Logger utility for Firebase Functions
 * Provides consistent logging with timestamps and environment-based debug logging
 */

// Environment check for debug logging
const isProduction = process.env.NODE_ENV === 'production';

/**
 * Format timestamp for log messages
 */
const getTimestamp = (): string => {
  return new Date().toISOString();
};

/**
 * Log levels and their corresponding console methods
 */
type LogLevel = 'log' | 'warn' | 'error' | 'debug';

/**
 * Core logging function with timestamp prefix
 */
const logWithLevel = (level: LogLevel, ...args: any[]): void => {
  const timestamp = getTimestamp();
  const prefix = `[${timestamp}]`;
  
  switch (level) {
    case 'log':
      console.log(prefix, ...args);
      break;
    case 'warn':
      console.warn(prefix, ...args);
      break;
    case 'error':
      console.error(prefix, ...args);
      break;
    case 'debug':
      if (!isProduction) {
        console.debug(prefix, ...args);
      }
      break;
  }
};

/**
 * Logger utility functions
 */
export const logger = {
  log: (...args: any[]) => logWithLevel('log', ...args),
  warn: (...args: any[]) => logWithLevel('warn', ...args),
  error: (...args: any[]) => logWithLevel('error', ...args),
  debug: (...args: any[]) => logWithLevel('debug', ...args),
};
