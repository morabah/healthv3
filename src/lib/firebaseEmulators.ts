import { getApp } from 'firebase/app';
import { connectAuthEmulator, getAuth } from 'firebase/auth';
import { connectFirestoreEmulator, getFirestore } from 'firebase/firestore';
import { connectFunctionsEmulator, getFunctions } from 'firebase/functions';
import { connectStorageEmulator, getStorage } from 'firebase/storage';

/**
 * Connect to Firebase emulators in development and test environments
 */
export const connectToEmulators = () => {
  try {
    // Only connect to emulators in development or test environment
    if (process.env.NODE_ENV === 'production') {
      return;
    }

    const app = getApp();
    
    // Auth emulator
    const auth = getAuth(app);
    connectAuthEmulator(auth, 'http://localhost:9099', { disableWarnings: true });
    
    // Firestore emulator
    const firestore = getFirestore(app);
    connectFirestoreEmulator(firestore, 'localhost', 8080);
    
    // Functions emulator
    const functions = getFunctions(app);
    connectFunctionsEmulator(functions, 'localhost', 5001);
    
    // Storage emulator
    const storage = getStorage(app);
    connectStorageEmulator(storage, 'localhost', 9199);
    
    console.log('Connected to Firebase emulators');
  } catch (error) {
    console.error('Error connecting to Firebase emulators:', error);
  }
};

/**
 * Helper function to check if emulators are running
 */
export const areEmulatorsRunning = async (): Promise<boolean> => {
  try {
    // Try to connect to the Emulator UI
    const response = await fetch('http://localhost:4000');
    return response.status === 200;
  } catch (error) {
    return false;
  }
};
