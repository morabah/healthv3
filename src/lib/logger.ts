/**
 * Logging utility for the Health Appointment System
 * Provides standardized logging functions with timestamps and log levels
 */

/**
 * Returns current ISO timestamp for consistent log formatting
 */
const getTimestamp = (): string => {
  return new Date().toISOString();
};

/**
 * Type definition for log parameters to ensure type safety
 */
export type LogParams = {
  message: string;
  data?: unknown;
  context?: string;
};

/**
 * Formats log message with timestamp, level, context, and optional data
 */
const formatLogMessage = (
  level: string,
  { message, data, context }: LogParams
): string => {
  const timestamp = getTimestamp();
  const contextStr = context ? `[${context}] ` : '';
  const baseMessage = `[${timestamp}] [${level}] ${contextStr}${message}`;
  
  if (data !== undefined) {
    return `${baseMessage} ${JSON.stringify(data, null, 2)}`;
  }
  
  return baseMessage;
};

/**
 * Logs informational messages
 * @param params Log parameters including message, optional data, and context
 */
export const logInfo = (params: LogParams): void => {
  console.log(formatLogMessage('INFO', params));
};

/**
 * Logs warning messages
 * @param params Log parameters including message, optional data, and context
 */
export const logWarn = (params: LogParams): void => {
  console.warn(formatLogMessage('WARN', params));
};

/**
 * Logs error messages
 * @param params Log parameters including message, optional data, and context
 */
export const logError = (params: LogParams): void => {
  console.error(formatLogMessage('ERROR', params));
};

/**
 * Logs debug messages only when NEXT_PUBLIC_LOG_LEVEL is set to 'debug'
 * @param params Log parameters including message, optional data, and context
 */
export const logDebug = (params: LogParams): void => {
  if (typeof process !== 'undefined' && 
      process.env.NEXT_PUBLIC_LOG_LEVEL === 'debug') {
    console.debug(formatLogMessage('DEBUG', params));
  }
};

/**
 * Default export for convenient importing
 */
export default {
  logInfo,
  logWarn,
  logError,
  logDebug,
};
