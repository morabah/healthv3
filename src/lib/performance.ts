/**
 * Performance tracking utility for the Health Appointment System
 * Provides a simple way to measure and log execution times
 */

import { logInfo } from './logger';

/**
 * Interface for the object returned by trackPerformance
 */
export interface PerformanceTracker {
  /**
   * Stops the performance tracking and logs the duration
   * @param additionalData Optional data to include in the log
   * @returns The duration in milliseconds
   */
  stop: (additionalData?: Record<string, unknown>) => number;
}

/**
 * Tracks the performance of an operation
 * @param label A descriptive label for the operation being tracked
 * @param context Optional context to include in the log (e.g., 'DatabaseService')
 * @returns An object with a stop() method to end tracking and log the duration
 */
export const trackPerformance = (
  label: string,
  context?: string
): PerformanceTracker => {
  // Use performance.now() for high-resolution timing
  const startTime = performance.now();
  
  return {
    stop: (additionalData?: Record<string, unknown>): number => {
      const endTime = performance.now();
      const duration = endTime - startTime;
      
      // Log the performance data using the logger
      logInfo({
        message: `${label} took ${duration.toFixed(2)} ms`,
        context: context || 'PERF',
        data: {
          duration,
          label,
          ...additionalData
        }
      });
      
      return duration;
    }
  };
};

/**
 * Default export for convenient importing
 */
export default {
  trackPerformance
};
