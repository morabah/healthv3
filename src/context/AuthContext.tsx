/**
 * Authentication Context
 * Provides authentication state and user profile data throughout the application
 */

import React, { createContext, useContext, useEffect, useState } from 'react';
import { User } from 'firebase/auth';
import { doc, getDoc } from 'firebase/firestore';
import { useAuthState } from 'react-firebase-hooks/auth';
import { getFirebaseAuth, getFirebaseFirestore } from '../lib/firebaseClient';
import { UserProfile } from '../types/user';
import { logInfo, logError } from '../lib/logger';
import { trackPerformance } from '../lib/performance';

// Context interface defining what data and functions will be available
interface AuthContextType {
  // Firebase authentication state
  user: User | null;
  loading: boolean;
  error: Error | undefined;
  
  // User profile data from Firestore
  userProfile: UserProfile | null;
  profileLoading: boolean;
  
  // Sign out function
  signOut: () => Promise<void>;
}

// Create the context with a default value
const AuthContext = createContext<AuthContextType>({
  user: null,
  loading: true,
  error: undefined,
  userProfile: null,
  profileLoading: true,
  signOut: async () => {},
});

// Custom hook to use the auth context
export const useAuth = () => useContext(AuthContext);

// Provider component that wraps the app and makes auth object available
export const AuthProvider: React.FC<{ children: React.ReactNode }> = ({ children }) => {
  // Get Firebase auth instance
  const auth = getFirebaseAuth();
  const firestore = getFirebaseFirestore();
  
  // Use the useAuthState hook to track authentication state
  const [user, loading, error] = useAuthState(auth);
  
  // State for user profile data from Firestore
  const [userProfile, setUserProfile] = useState<UserProfile | null>(null);
  const [profileLoading, setProfileLoading] = useState<boolean>(true);
  
  // Function to sign out
  const signOut = async () => {
    try {
      logInfo({
        message: 'User signing out',
        context: 'AuthContext',
        data: { userId: user?.uid }
      });
      
      await auth.signOut();
      
      // Clear user profile state
      setUserProfile(null);
      
      logInfo({
        message: 'User signed out successfully',
        context: 'AuthContext'
      });
    } catch (error) {
      logError({
        message: 'Error signing out',
        context: 'AuthContext',
        data: { error }
      });
      throw error;
    }
  };
  
  // Effect to fetch user profile data when auth state changes
  useEffect(() => {
    // Skip if still loading or no user
    if (loading || !user) {
      if (!loading && !user) {
        // User is definitely signed out
        setProfileLoading(false);
        setUserProfile(null);
        
        logInfo({
          message: 'No authenticated user',
          context: 'AuthContext'
        });
      }
      return;
    }
    
    const fetchUserProfile = async () => {
      // Start performance tracking
      const perfTracker = trackPerformance('fetchUserProfile', 'AuthContext');
      
      try {
        setProfileLoading(true);
        
        logInfo({
          message: 'Fetching user profile',
          context: 'AuthContext',
          data: { userId: user.uid }
        });
        
        // Get user document from Firestore
        const userDocRef = doc(firestore, 'users', user.uid);
        const userDoc = await getDoc(userDocRef);
        
        if (userDoc.exists()) {
          // User profile exists in Firestore
          const userData = userDoc.data() as UserProfile;
          setUserProfile(userData);
          
          logInfo({
            message: 'User profile fetched successfully',
            context: 'AuthContext',
            data: { userId: user.uid, userType: userData.userType }
          });
        } else {
          // User authenticated but no profile in Firestore
          setUserProfile(null);
          
          logInfo({
            message: 'User authenticated but no profile found',
            context: 'AuthContext',
            data: { userId: user.uid }
          });
        }
      } catch (error) {
        logError({
          message: 'Error fetching user profile',
          context: 'AuthContext',
          data: { userId: user.uid, error }
        });
        
        setUserProfile(null);
      } finally {
        setProfileLoading(false);
        perfTracker.stop({ userId: user.uid });
      }
    };
    
    // Log authentication state change
    logInfo({
      message: 'User authenticated',
      context: 'AuthContext',
      data: { 
        userId: user.uid,
        email: user.email,
        emailVerified: user.emailVerified,
        isAnonymous: user.isAnonymous
      }
    });
    
    // Fetch the user profile
    fetchUserProfile();
  }, [user, loading, firestore]);
  
  // Create the value object that will be provided by the context
  const value = {
    user,
    loading,
    error,
    userProfile,
    profileLoading,
    signOut
  };
  
  return (
    <AuthContext.Provider value={value}>
      {children}
    </AuthContext.Provider>
  );
};

export default AuthContext;
