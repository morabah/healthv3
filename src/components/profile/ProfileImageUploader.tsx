/**
 * ProfileImageUploader Component
 * Allows users to upload and manage their profile image
 */
import React, { useState, useRef } from 'react';
import { getStorage, ref, uploadBytesResumable, getDownloadURL, deleteObject } from 'firebase/storage';
import { useAuth } from '@/context/AuthContext';
import { logInfo, logError } from '@/lib/logger';
import { trackPerformance } from '@/lib/performance';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faUser, faUpload, faSpinner, faTrash } from '@fortawesome/free-solid-svg-icons';
import Button from '../ui/Button';

interface ProfileImageUploaderProps {
  currentImageUrl?: string;
  onImageUploaded: (imageUrl: string) => void;
  onImageDeleted?: () => void;
}

const ProfileImageUploader: React.FC<ProfileImageUploaderProps> = ({
  currentImageUrl,
  onImageUploaded,
  onImageDeleted
}) => {
  const { user } = useAuth();
  const [uploading, setUploading] = useState<boolean>(false);
  const [uploadProgress, setUploadProgress] = useState<number>(0);
  const [error, setError] = useState<string | null>(null);
  const fileInputRef = useRef<HTMLInputElement>(null);
  
  const handleFileSelect = () => {
    if (fileInputRef.current) {
      fileInputRef.current.click();
    }
  };
  
  const handleFileChange = async (e: React.ChangeEvent<HTMLInputElement>) => {
    const file = e.target.files?.[0];
    if (!file || !user) return;
    
    // Validate file type
    if (!file.type.match('image.*')) {
      setError('Please select an image file (PNG, JPG, JPEG)');
      return;
    }
    
    // Validate file size (max 5MB)
    if (file.size > 5 * 1024 * 1024) {
      setError('Image size should be less than 5MB');
      return;
    }
    
    setUploading(true);
    setUploadProgress(0);
    setError(null);
    
    const perfTracker = trackPerformance('uploadProfileImage', 'ProfileImageUploader');
    
    try {
      logInfo({
        message: 'Starting profile image upload',
        context: 'ProfileImageUploader',
        data: { userId: user.uid, fileSize: file.size, fileType: file.type }
      });
      
      const storage = getStorage();
      const storageRef = ref(storage, `profile-images/${user.uid}`);
      
      const uploadTask = uploadBytesResumable(storageRef, file);
      
      // Listen for upload progress
      uploadTask.on(
        'state_changed',
        (snapshot) => {
          const progress = Math.round(
            (snapshot.bytesTransferred / snapshot.totalBytes) * 100
          );
          setUploadProgress(progress);
        },
        (error) => {
          logError({
            message: 'Profile image upload failed',
            context: 'ProfileImageUploader',
            data: { userId: user.uid, error }
          });
          
          setError('Failed to upload image. Please try again.');
          setUploading(false);
          perfTracker.stop({ error: true });
        },
        async () => {
          // Upload completed successfully
          const downloadUrl = await getDownloadURL(uploadTask.snapshot.ref);
          
          logInfo({
            message: 'Profile image uploaded successfully',
            context: 'ProfileImageUploader',
            data: { userId: user.uid, downloadUrl }
          });
          
          onImageUploaded(downloadUrl);
          setUploading(false);
          perfTracker.stop({ success: true });
        }
      );
    } catch (error: any) {
      logError({
        message: 'Error in profile image upload',
        context: 'ProfileImageUploader',
        data: { userId: user.uid, error }
      });
      
      setError(error.message || 'An unexpected error occurred');
      setUploading(false);
      perfTracker.stop({ error: true });
    }
  };
  
  const handleDeleteImage = async () => {
    if (!user || !currentImageUrl) return;
    
    setUploading(true);
    setError(null);
    
    const perfTracker = trackPerformance('deleteProfileImage', 'ProfileImageUploader');
    
    try {
      logInfo({
        message: 'Deleting profile image',
        context: 'ProfileImageUploader',
        data: { userId: user.uid, imageUrl: currentImageUrl }
      });
      
      const storage = getStorage();
      const storageRef = ref(storage, `profile-images/${user.uid}`);
      
      await deleteObject(storageRef);
      
      logInfo({
        message: 'Profile image deleted successfully',
        context: 'ProfileImageUploader',
        data: { userId: user.uid }
      });
      
      if (onImageDeleted) {
        onImageDeleted();
      }
      
      perfTracker.stop({ success: true });
    } catch (error: any) {
      logError({
        message: 'Error deleting profile image',
        context: 'ProfileImageUploader',
        data: { userId: user.uid, error }
      });
      
      setError(error.message || 'Failed to delete image');
      perfTracker.stop({ error: true });
    } finally {
      setUploading(false);
    }
  };
  
  return (
    <div className="profile-image-uploader" data-testid="profile-image-uploader">
      <div className="flex items-center space-x-4">
        <div className="relative">
          <div className="w-24 h-24 rounded-full overflow-hidden bg-gray-200 flex items-center justify-center">
            {currentImageUrl ? (
              <img 
                src={currentImageUrl} 
                alt="Profile" 
                className="w-full h-full object-cover"
                data-testid="profile-image"
              />
            ) : (
              <FontAwesomeIcon 
                icon={faUser} 
                className="text-gray-400 text-4xl"
                data-testid="default-icon"
              />
            )}
            
            {uploading && (
              <div className="absolute inset-0 bg-black bg-opacity-50 flex items-center justify-center rounded-full">
                <div className="text-center">
                  <FontAwesomeIcon icon={faSpinner} spin className="text-white text-xl" />
                  <div className="text-white text-xs mt-1">{uploadProgress}%</div>
                </div>
              </div>
            )}
          </div>
        </div>
        
        <div className="flex flex-col space-y-2">
          <input
            type="file"
            ref={fileInputRef}
            className="hidden"
            accept="image/*"
            onChange={handleFileChange}
            data-testid="file-input"
          />
          
          <Button
            variant="primary"
            size="sm"
            onClick={handleFileSelect}
            disabled={uploading}
            icon={faUpload}
            data-testid="upload-button"
          >
            {currentImageUrl ? 'Change Photo' : 'Upload Photo'}
          </Button>
          
          {currentImageUrl && (
            <Button
              variant="danger"
              size="sm"
              onClick={handleDeleteImage}
              disabled={uploading}
              icon={faTrash}
              data-testid="delete-button"
            >
              Remove Photo
            </Button>
          )}
        </div>
      </div>
      
      {error && (
        <div className="mt-2 text-sm text-red-600" data-testid="error-message">
          {error}
        </div>
      )}
      
      <div className="mt-2 text-xs text-gray-500">
        Supported formats: JPG, PNG. Max size: 5MB
      </div>
    </div>
  );
};

export default ProfileImageUploader;
