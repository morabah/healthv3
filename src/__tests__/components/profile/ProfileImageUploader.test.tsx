/**
 * ProfileImageUploader Component Tests
 * Tests the profile image uploader component with mocked Firebase Storage
 */
import React from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import '@testing-library/jest-dom';
import ProfileImageUploader from '@/components/profile/ProfileImageUploader';
import { useAuth } from '@/context/AuthContext';
import { getStorage, ref, uploadBytesResumable, getDownloadURL, deleteObject } from 'firebase/storage';
import { logInfo, logError } from '@/lib/logger';
import { trackPerformance } from '@/lib/performance';

// Mock Firebase Storage
jest.mock('firebase/storage', () => ({
  getStorage: jest.fn(),
  ref: jest.fn(),
  uploadBytesResumable: jest.fn(),
  getDownloadURL: jest.fn(),
  deleteObject: jest.fn(),
}));

// Mock Auth context
jest.mock('@/context/AuthContext', () => ({
  useAuth: jest.fn(),
}));

// Mock logger and performance utilities
jest.mock('@/lib/logger', () => ({
  logInfo: jest.fn(),
  logError: jest.fn(),
}));

jest.mock('@/lib/performance', () => ({
  trackPerformance: jest.fn(() => ({
    stop: jest.fn(),
  })),
}));

describe('ProfileImageUploader Component', () => {
  // Mock functions
  const mockGetStorage = getStorage as jest.Mock;
  const mockRef = ref as jest.Mock;
  const mockUploadBytesResumable = uploadBytesResumable as jest.Mock;
  const mockGetDownloadURL = getDownloadURL as jest.Mock;
  const mockDeleteObject = deleteObject as jest.Mock;
  const mockUseAuth = useAuth as jest.Mock;
  const mockOnImageUploaded = jest.fn();
  const mockOnImageDeleted = jest.fn();
  
  // Default props
  const defaultProps = {
    onImageUploaded: mockOnImageUploaded,
    onImageDeleted: mockOnImageDeleted,
  };
  
  // Mock file
  const createMockFile = (type = 'image/jpeg', size = 1024 * 1024) => {
    const file = new File(['mock-content'], 'profile.jpg', { type });
    Object.defineProperty(file, 'size', { value: size });
    return file;
  };
  
  beforeEach(() => {
    jest.clearAllMocks();
    
    // Mock authenticated user
    mockUseAuth.mockReturnValue({
      user: { uid: 'user123' }
    });
    
    // Mock storage functions
    mockGetStorage.mockReturnValue({});
    mockRef.mockReturnValue('storageRef');
    
    // Mock upload task
    const mockUploadTask = {
      on: jest.fn((event, progressCallback, errorCallback, completeCallback) => {
        // Simulate successful upload by default
        progressCallback({ bytesTransferred: 100, totalBytes: 100 });
        completeCallback();
      }),
      snapshot: {
        ref: 'uploadTaskRef'
      }
    };
    
    mockUploadBytesResumable.mockReturnValue(mockUploadTask);
    mockGetDownloadURL.mockResolvedValue('https://example.com/profile.jpg');
    mockDeleteObject.mockResolvedValue(undefined);
  });
  
  it('renders without a current image', () => {
    render(<ProfileImageUploader {...defaultProps} />);
    
    // Should show default icon
    expect(screen.getByTestId('default-icon')).toBeInTheDocument();
    
    // Should show upload button with correct text
    expect(screen.getByTestId('upload-button')).toHaveTextContent('Upload Photo');
    
    // Should not show delete button
    expect(screen.queryByTestId('delete-button')).not.toBeInTheDocument();
  });
  
  it('renders with a current image', () => {
    render(
      <ProfileImageUploader 
        {...defaultProps} 
        currentImageUrl="https://example.com/existing.jpg" 
      />
    );
    
    // Should show the image
    const profileImage = screen.getByTestId('profile-image');
    expect(profileImage).toBeInTheDocument();
    expect(profileImage).toHaveAttribute('src', 'https://example.com/existing.jpg');
    
    // Should show upload button with "Change Photo" text
    expect(screen.getByTestId('upload-button')).toHaveTextContent('Change Photo');
    
    // Should show delete button
    expect(screen.getByTestId('delete-button')).toBeInTheDocument();
  });
  
  it('opens file dialog when upload button is clicked', async () => {
    render(<ProfileImageUploader {...defaultProps} />);
    
    const fileInput = screen.getByTestId('file-input');
    const uploadButton = screen.getByTestId('upload-button');
    
    // Mock the click function of the file input
    const clickSpy = jest.spyOn(fileInput, 'click');
    
    // Click the upload button
    await userEvent.click(uploadButton);
    
    // Verify file input click was triggered
    expect(clickSpy).toHaveBeenCalled();
  });
  
  it('uploads an image file successfully', async () => {
    render(<ProfileImageUploader {...defaultProps} />);
    
    const fileInput = screen.getByTestId('file-input');
    const file = createMockFile();
    
    // Simulate file selection
    await userEvent.upload(fileInput, file);
    
    // Verify storage functions were called correctly
    expect(mockGetStorage).toHaveBeenCalled();
    expect(mockRef).toHaveBeenCalledWith({}, 'profile-images/user123');
    expect(mockUploadBytesResumable).toHaveBeenCalledWith('storageRef', file);
    
    // Verify download URL was fetched
    expect(mockGetDownloadURL).toHaveBeenCalledWith('uploadTaskRef');
    
    // Verify callback was called with the download URL
    await waitFor(() => {
      expect(mockOnImageUploaded).toHaveBeenCalledWith('https://example.com/profile.jpg');
    });
    
    // Verify logging
    expect(logInfo).toHaveBeenCalledWith(
      expect.objectContaining({
        message: 'Starting profile image upload',
        context: 'ProfileImageUploader'
      })
    );
    
    expect(logInfo).toHaveBeenCalledWith(
      expect.objectContaining({
        message: 'Profile image uploaded successfully',
        context: 'ProfileImageUploader'
      })
    );
  });
  
  it('shows upload progress', async () => {
    // Mock upload task with progress updates
    const progressUpdates = [
      { bytesTransferred: 25, totalBytes: 100 }, // 25%
      { bytesTransferred: 50, totalBytes: 100 }, // 50%
      { bytesTransferred: 75, totalBytes: 100 }, // 75%
      { bytesTransferred: 100, totalBytes: 100 } // 100%
    ];
    
    let progressCallback: any;
    let completeCallback: any;
    
    mockUploadBytesResumable.mockReturnValue({
      on: jest.fn((event, onProgress, onError, onComplete) => {
        progressCallback = onProgress;
        completeCallback = onComplete;
      }),
      snapshot: {
        ref: 'uploadTaskRef'
      }
    });
    
    render(<ProfileImageUploader {...defaultProps} />);
    
    const fileInput = screen.getByTestId('file-input');
    const file = createMockFile();
    
    // Simulate file selection
    await userEvent.upload(fileInput, file);
    
    // Simulate progress updates
    progressCallback(progressUpdates[0]);
    
    // Check for progress indicator
    await waitFor(() => {
      const progressText = screen.getByText('25%');
      expect(progressText).toBeInTheDocument();
    });
    
    // Complete the upload
    completeCallback();
    
    // Verify callback was called
    await waitFor(() => {
      expect(mockOnImageUploaded).toHaveBeenCalled();
    });
  });
  
  it('handles non-image file upload attempts', async () => {
    render(<ProfileImageUploader {...defaultProps} />);
    
    const fileInput = screen.getByTestId('file-input');
    const file = new File(['mock-content'], 'document.pdf', { type: 'application/pdf' });
    
    // Simulate file selection
    await userEvent.upload(fileInput, file);
    
    // Verify error message
    await waitFor(() => {
      expect(screen.getByTestId('error-message')).toHaveTextContent('Please select an image file (PNG, JPG, JPEG)');
    });
    
    // Verify storage functions were not called
    expect(mockUploadBytesResumable).not.toHaveBeenCalled();
  });
  
  it('handles oversized file upload attempts', async () => {
    render(<ProfileImageUploader {...defaultProps} />);
    
    const fileInput = screen.getByTestId('file-input');
    const file = createMockFile('image/jpeg', 6 * 1024 * 1024); // 6MB file (over the 5MB limit)
    
    // Simulate file selection
    await userEvent.upload(fileInput, file);
    
    // Verify error message
    await waitFor(() => {
      expect(screen.getByTestId('error-message')).toHaveTextContent('Image size should be less than 5MB');
    });
    
    // Verify storage functions were not called
    expect(mockUploadBytesResumable).not.toHaveBeenCalled();
  });
  
  it('handles upload errors', async () => {
    // Mock upload task with error
    const mockError = new Error('Network error');
    
    let errorCallback: any;
    
    mockUploadBytesResumable.mockReturnValue({
      on: jest.fn((event, onProgress, onError) => {
        errorCallback = onError;
      }),
      snapshot: {
        ref: 'uploadTaskRef'
      }
    });
    
    render(<ProfileImageUploader {...defaultProps} />);
    
    const fileInput = screen.getByTestId('file-input');
    const file = createMockFile();
    
    // Simulate file selection
    await userEvent.upload(fileInput, file);
    
    // Simulate upload error
    errorCallback(mockError);
    
    // Verify error message
    await waitFor(() => {
      expect(screen.getByTestId('error-message')).toHaveTextContent('Failed to upload image. Please try again.');
    });
    
    // Verify error logging
    expect(logError).toHaveBeenCalledWith(
      expect.objectContaining({
        message: 'Profile image upload failed',
        context: 'ProfileImageUploader',
        data: expect.objectContaining({
          userId: 'user123',
          error: mockError
        })
      })
    );
    
    // Verify callback was not called
    expect(mockOnImageUploaded).not.toHaveBeenCalled();
  });
  
  it('deletes an existing image', async () => {
    render(
      <ProfileImageUploader 
        {...defaultProps} 
        currentImageUrl="https://example.com/existing.jpg" 
      />
    );
    
    const deleteButton = screen.getByTestId('delete-button');
    
    // Click delete button
    await userEvent.click(deleteButton);
    
    // Verify storage functions were called correctly
    expect(mockGetStorage).toHaveBeenCalled();
    expect(mockRef).toHaveBeenCalledWith({}, 'profile-images/user123');
    expect(mockDeleteObject).toHaveBeenCalledWith('storageRef');
    
    // Verify callback was called
    await waitFor(() => {
      expect(mockOnImageDeleted).toHaveBeenCalled();
    });
    
    // Verify logging
    expect(logInfo).toHaveBeenCalledWith(
      expect.objectContaining({
        message: 'Deleting profile image',
        context: 'ProfileImageUploader'
      })
    );
    
    expect(logInfo).toHaveBeenCalledWith(
      expect.objectContaining({
        message: 'Profile image deleted successfully',
        context: 'ProfileImageUploader'
      })
    );
  });
  
  it('handles delete errors', async () => {
    // Mock delete error
    const mockError = new Error('Permission denied');
    mockDeleteObject.mockRejectedValueOnce(mockError);
    
    render(
      <ProfileImageUploader 
        {...defaultProps} 
        currentImageUrl="https://example.com/existing.jpg" 
      />
    );
    
    const deleteButton = screen.getByTestId('delete-button');
    
    // Click delete button
    await userEvent.click(deleteButton);
    
    // Verify error message
    await waitFor(() => {
      expect(screen.getByTestId('error-message')).toHaveTextContent('Permission denied');
    });
    
    // Verify error logging
    expect(logError).toHaveBeenCalledWith(
      expect.objectContaining({
        message: 'Error deleting profile image',
        context: 'ProfileImageUploader',
        data: expect.objectContaining({
          userId: 'user123',
          error: mockError
        })
      })
    );
    
    // Verify callback was not called
    expect(mockOnImageDeleted).not.toHaveBeenCalled();
  });
  
  it('handles missing user', async () => {
    // Mock unauthenticated state
    mockUseAuth.mockReturnValue({
      user: null
    });
    
    render(
      <ProfileImageUploader 
        {...defaultProps} 
        currentImageUrl="https://example.com/existing.jpg" 
      />
    );
    
    const fileInput = screen.getByTestId('file-input');
    const file = createMockFile();
    
    // Simulate file selection
    await userEvent.upload(fileInput, file);
    
    // Verify storage functions were not called
    expect(mockUploadBytesResumable).not.toHaveBeenCalled();
    
    // Click delete button
    const deleteButton = screen.getByTestId('delete-button');
    await userEvent.click(deleteButton);
    
    // Verify delete was not called
    expect(mockDeleteObject).not.toHaveBeenCalled();
  });
  
  it('matches snapshot', () => {
    const { container } = render(<ProfileImageUploader {...defaultProps} />);
    expect(container).toMatchSnapshot();
  });
});
