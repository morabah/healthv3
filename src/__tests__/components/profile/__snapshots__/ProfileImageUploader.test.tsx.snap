// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfileImageUploader Component matches snapshot 1`] = `
<div>
  <div
    class="profile-image-uploader"
    data-testid="profile-image-uploader"
  >
    <div
      class="flex items-center space-x-4"
    >
      <div
        class="relative"
      >
        <div
          class="w-24 h-24 rounded-full overflow-hidden bg-gray-200 flex items-center justify-center"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-user text-gray-400 text-4xl"
            data-icon="user"
            data-prefix="fas"
            data-testid="default-icon"
            focusable="false"
            role="img"
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M224 256A128 128 0 1 0 224 0a128 128 0 1 0 0 256zm-45.7 48C79.8 304 0 383.8 0 482.3C0 498.7 13.3 512 29.7 512l388.6 0c16.4 0 29.7-13.3 29.7-29.7C448 383.8 368.2 304 269.7 304l-91.4 0z"
              fill="currentColor"
            />
          </svg>
        </div>
      </div>
      <div
        class="flex flex-col space-y-2"
      >
        <input
          accept="image/*"
          class="hidden"
          data-testid="file-input"
          type="file"
        />
        <button
          class="
    inline-flex items-center justify-center font-medium transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2
    text-sm py-1.5 px-3
    bg-blue-600 hover:bg-blue-700 text-white focus:ring-blue-500 bg-blue
    
    
    rounded-md
    
  "
          data-testid="upload-button"
        >
          <svg
            aria-hidden="true"
            class="svg-inline--fa fa-upload mr-2"
            data-icon="upload"
            data-prefix="fas"
            focusable="false"
            role="img"
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M288 109.3L288 352c0 17.7-14.3 32-32 32s-32-14.3-32-32l0-242.7-73.4 73.4c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3l128-128c12.5-12.5 32.8-12.5 45.3 0l128 128c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L288 109.3zM64 352l128 0c0 35.3 28.7 64 64 64s64-28.7 64-64l128 0c35.3 0 64 28.7 64 64l0 32c0 35.3-28.7 64-64 64L64 512c-35.3 0-64-28.7-64-64l0-32c0-35.3 28.7-64 64-64zM432 456a24 24 0 1 0 0-48 24 24 0 1 0 0 48z"
              fill="currentColor"
            />
          </svg>
          Upload Photo
        </button>
      </div>
    </div>
    <div
      class="mt-2 text-xs text-gray-500"
    >
      Supported formats: JPG, PNG. Max size: 5MB
    </div>
  </div>
</div>
`;
